resilience4j-rxjava2/src/main/java/io/github/resilience4j/retry/transformer/RetryTransformer.java

```diff
@@ -58,51 +53,67 @@ private RetryTransformer(Retry retry) {
 
     @Override
     public Publisher<T> apply(Flowable<T> upstream) {
-        return Flowable.fromPublisher(downstream -> {
-            SubscriptionArbiter sa = new SubscriptionArbiter();
-            downstream.onSubscribe(sa);
-            RetrySubscriber<T> repeatSubscriber = new RetrySubscriber<>(downstream, retry.getRetryConfig().getMaxAttempts(), sa, upstream, retry);
-            upstream.subscribe(repeatSubscriber);
-        });
+        return Flowable.fromPublisher(downstream -> applyRetrySubscriber(downstream, upstream));
     }
 
     @Override
     public ObservableSource<T> apply(Observable<T> upstream) {
-        return Observable.fromPublisher(downstream -> {
-            Flowable<T> flowable = upstream.toFlowable(BackpressureStrategy.BUFFER);
-            SubscriptionArbiter sa = new SubscriptionArbiter();
-            downstream.onSubscribe(sa);
-            RetrySubscriber<T> retrySubscriber = new RetrySubscriber<>(downstream, retry.getRetryConfig().getMaxAttempts(), sa, flowable, retry);
-            flowable.subscribe(retrySubscriber);
-        });
+        return Observable.fromPublisher(downstream ->
+                applyRetrySubscriber(downstream, upstream.toFlowable(BackpressureStrategy.BUFFER)));
     }
 
     @Override
     public SingleSource<T> apply(Single<T> upstream) {
-        return Single.fromPublisher(downstream -> {
+        return Single.fromPublisher(downstream -> applyRetrySubscriber(downstream, upstream.toFlowable()));
+    }
+
+    @Override
+    public CompletableSource apply(Completable upstream) {
+        return Completable.fromPublisher(downstream -> {
             Flowable<T> flowable = upstream.toFlowable();
-            SubscriptionArbiter sa = new SubscriptionArbiter();
+            SubscriptionArbiter sa = new SubscriptionArbiter(true);
             downstream.onSubscribe(sa);
-            RetrySubscriber<T> retrySubscriber = new RetrySubscriber<>(downstream, retry.getRetryConfig().getMaxAttempts(), sa, flowable, retry);
+            //noinspection unchecked
+            RetrySubscriber<T> retrySubscriber = new RetrySubscriber<>(downstream,
+                    retry.getRetryConfig().getMaxAttempts(), sa, flowable, retry.context(), true);
             flowable.subscribe(retrySubscriber);
         });
     }
 
+    @SuppressWarnings("unchecked")
+    private void applyRetrySubscriber(Subscriber<? super T> downstream, Flowable<T> flowable) {
+        Retry.Context<T> context = retry.context();
+        flowable = flowable.doOnNext(value -> throwExceptionToForceRetryOnResult(context, value));
```

Comment:
This change uses an Exception to change the flow control.
Could we instead invoke `context.onResult(...)` in  `RetrySubscriber.onNext(...)` and invoke `subscribeNext` when onResult returns true?
Or doesn't this work?

Reply:
Hi @RobWin,
I even have that in my local history :smile:. In short it does not work, but let me elaborate on that a little bit.

When we do that, this is what happens:
1. `RetrySubscriber.onSubscribe()` is called for the first time
2. `RetrySubscriber.onNext()` receives the value, checks for the `context.onResult(t)` and proceeds to `subscribeNext()`
3. `subscribeNext()` gets called and passes the trampolining check since it's our first "recurrent" call
    1.`RetrySubscriber.onSubscribe()` is called for the second time (our retry)
    2. `SubscriptionArbiter.setSubscription` is called and our previous subscription is cancelled, so when it finishes at some point in time it won't call `onComplete()`. Lastly our new stream starts
    3. `RetrySubscriber.onNext` is called, but this time since we are in recurrent call trampolining won't allow doing subscribe again in this call stack
    4. `RetrySubscriber.onNext` will finish, but since we didn't call the subscribe immediately the stream was not closed and RxJava will notify the subscriber that the stream completed. At this point we are already broken since there will be at least few more `onComplete` calls for the subscriptions scheduled via trampolining.

By the way I came up with another approach that seems to work and we do not have to user RxJava's internal api's. I'll push that any minute

Reply:
Thanks for your answer